common/scala/src/main/scala/org/apache/openwhisk/core/yarn/YARNContainerFactory.scala

```diff
@@ -0,0 +1,476 @@
+    serviceLock.synchronized {
+
+      val componentList = images
+        .map(
+          i =>
+            ComponentDefinition(
+              i.name.replace('.', '-'), //name must be [a-z][a-z0-9-]*
+              0, //start with zero containers
+              runCommand,
+              Option.empty,
+              ArtifactDefinition(i.publicImageName, "DOCKER"),
+              ResourceDefinition(yarnConfig.cpus, yarnConfig.memory),
+              ConfigurationDefinition(Map(("YARN_CONTAINER_RUNTIME_DOCKER_RUN_OVERRIDE_DISABLE", "true")))))
+        .toList
+
+      //Add kerberos def if necessary
+      var kerberosDef: Option[KerberosPrincipalDefinition] = None
+      if (yarnConfig.authType.equals(YARNContainerFactory.KERBEROSAUTH))
+        kerberosDef =
+          Some(KerberosPrincipalDefinition(Some(yarnConfig.kerberosPrincipal), Some(yarnConfig.kerberosKeytab)))
+
+      val service = ServiceDefinition(
+        yarnConfig.serviceName,
+        Some(version),
+        Some(description),
+        "STABLE",
+        Some(yarnConfig.queue),
+        componentList,
+        kerberosDef)
+
+      //Submit service
+      val response =
+        submitRequestWithAuth(HttpMethods.POST, s"${yarnConfig.masterUrl}/app/v1/services", service.toJson.compactPrint)
+
+      //Handle response
+      response match {
+        case httpresponse(StatusCodes.OK, content) =>
+          logging.info(this, s"Service submitted. Response: $content")
+
+        case httpresponse(StatusCodes.Accepted, content) =>
+          logging.info(this, s"Service submitted. Response: $content")
+
+        case httpresponse(_, _) => handleYARNRESTError(response)
+      }
+    }
+  }
+
+  override def createContainer(
+    tid: TransactionId,
+    name: String,
+    actionImage: ExecManifest.ImageName,
+    unuseduserProvidedImage: Boolean,
+    unusedmemory: ByteSize,
+    unusedcpuShares: Int)(implicit config: WhiskConfig, logging: Logging): Future[Container] = {
+
+    logging.info(this, s"Using YARN to create a container with image ${actionImage.name}...")
+
+    Future {
+      serviceLock.synchronized {
+        val newContainerCount = containerCounts.getOrElse(actionImage.name, 0) + 1
+        val body = "{\"number_of_containers\":" + newContainerCount + "}"
+        val response = submitRequestWithAuth(
+          HttpMethods.PUT,
+          s"${yarnConfig.masterUrl}/app/v1/services/${yarnConfig.serviceName}/components/${actionImage.name}",
+          body)
+        response match {
+          case httpresponse(StatusCodes.OK, content) =>
+            logging.info(this, s"Added container: ${actionImage.name}. Response: $content")
+            containerCounts(actionImage.name) = newContainerCount
+
+          case httpresponse(_, _) => handleYARNRESTError(response)
+        }
+
+        //Wait for container to come up
+        val serviceDef = waitForContainerStability(actionImage)
```

Comment:
this suggest more blocking inside the future - if this is really necessary, there's a way to signal to the future scheduler that it can context switch here.

but i wonder if you should instead be chaining the futures (will come back to this after I've understood the rest of the code).